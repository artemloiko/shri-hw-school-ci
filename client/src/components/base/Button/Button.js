import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@reach/router';
import { cn } from 'utils/bem-cn';

import './Button.css';

function Button(props) {
  const { type, to = '', children, icon } = props;

  return type ? (
    <button type={type} className={cn('button', props)}>
      {icon}
      <div className="button__text">{children}</div>
    </button>
  ) : (
    <Link to={to} className={cn('button', props)}>
      {icon}
      <div className="button__text">{children}</div>
    </Link>
  );
}

Button.propTypes = {
  type: PropTypes.string,
  to: PropTypes.string,
  children: PropTypes.string,
  icon: PropTypes.element,
  className: PropTypes.string,
  mods: PropTypes.shape({
    size: PropTypes.oneOf(['small']),
    icon: PropTypes.bool,
    action: PropTypes.bool,
    'mini-desktop': PropTypes.bool,
  }),
  mix: PropTypes.arrayOf(PropTypes.string),
};

export default Button;
