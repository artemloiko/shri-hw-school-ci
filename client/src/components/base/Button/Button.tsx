import React from 'react';
import { Link } from '@reach/router';
import { cn, CNProps } from 'utils/bem-cn';

import Icon, { IconTypes } from 'components/base/Icon/Icon';

import './Button.css';

export type ButtonMods = {
  size?: 'small';
  icon?: boolean;
  action?: boolean;
  'mini-desktop'?: boolean;
  'icon-only'?: boolean;
  disabled?: boolean;
};

export interface ButtonProps {
  type?: 'submit' | 'reset' | 'button' | undefined;
  to?: string;
  children: string;
  iconType: IconTypes;
}

const Button: React.FC<ButtonProps & CNProps<ButtonMods>> = (props) => {
  const {
    type = 'button',
    to = '',
    children,
    iconType,
    mods = {},
    className,
    mix,
    ...buttonProps
  } = props;
  if (iconType) mods.icon = true;

  return to ? (
    <Link
      {...buttonProps}
      to={to}
      className={cn('button', { className, mods, mix })}
      tabIndex={mods.disabled ? -1 : 0}
    >
      {iconType && <Icon mods={{ size: 'small', type: iconType }} mix={['button']} />}
      <div className="button__text">{children}</div>
    </Link>
  ) : (
    <button
      {...buttonProps}
      type={type}
      className={cn('button', { className, mods, mix })}
      disabled={mods.disabled}
    >
      {iconType && <Icon mods={{ size: 'small', type: iconType }} mix={['button']} />}
      <div className="button__text">{children}</div>
    </button>
  );
};

export default Button;
