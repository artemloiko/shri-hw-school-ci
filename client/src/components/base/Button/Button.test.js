import React from 'react';
import { create } from 'react-test-renderer';
import { render } from '@testing-library/react';

import Button from './Button';

test('Button with default props matches the snapshot', () => {
  const tree = create(<Button>Default button</Button>).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Button[type=link] with all props matches the snapshot', () => {
  const tree = create(
    <Button iconType="play" className="block" mods={{ action: true }} to="/" mix={['block']}>
      Link
    </Button>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Button[type=submit] with all props matches the snapshot', () => {
  const tree = create(
    <Button iconType="play" className="block" mods={{ action: true }} type="submit" mix={['block']}>
      Button
    </Button>,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Button[disabled] is not focusable or clickable', () => {
  const { getByTestId, rerender } = render(
    <Button mods={{ disabled: true }} to="/" data-testid="btn">
      Link
    </Button>,
  );

  expect(getByTestId('btn')).toHaveAttribute('tabindex', '-1');
  expect(getByTestId('btn')).toHaveClass('button_disabled');

  rerender(
    <Button mods={{ disabled: true }} data-testid="btn">
      Link
    </Button>,
  );
  expect(getByTestId('btn')).toBeDisabled();
});
