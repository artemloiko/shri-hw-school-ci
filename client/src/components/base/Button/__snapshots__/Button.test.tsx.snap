// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button with default props matches the snapshot 1`] = `
<button
  className="button"
>
  <div
    className="button__text"
  >
    Default button
  </div>
</button>
`;

exports[`Button[type=link] with all props matches the snapshot 1`] = `
<a
  aria-current="page"
  className="block button button_action button_icon block__button"
  href="/"
  onClick={[Function]}
  tabIndex={0}
>
  <div
    className="icon icon_size_small icon_type_play button__icon"
  />
  <div
    className="button__text"
  >
    Link
  </div>
</a>
`;

exports[`Button[type=submit] with all props matches the snapshot 1`] = `
<button
  className="block button button_action button_icon block__button"
  type="submit"
>
  <div
    className="icon icon_size_small icon_type_play button__icon"
  />
  <div
    className="button__text"
  >
    Button
  </div>
</button>
`;
